fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvPartitionCompensator.java

```diff
@@ -107,104 +116,328 @@ public class MvPartitionCompensator {
+        } catch (Exception e) {
+            return null;
+        }
+        if (scanOperatorPredicates == null) {
+            return null;
+        }
+        List<PartitionKey> refTableCompensatePartitionKeys = Lists.newArrayList();
+        List<PartitionKey> selectPartitionKeys = scanOperatorPredicates.getSelectedPartitionKeys();
+        // different behavior for different external table types
+        if (selectPartitionKeys.isEmpty() && refScanOperator.getOpType() != OperatorType.LOGICAL_HIVE_SCAN) {
+            return null;
+        }
+        for (PartitionKey partitionKey : selectPartitionKeys) {
+            String partitionName = generateMVPartitionName(partitionKey);
+            if (refTablePartitionNamesToRefresh.contains(partitionName)) {
+                refTableCompensatePartitionKeys.add(partitionKey);
+            }
+        }
+        return refTableCompensatePartitionKeys;
+    }
+    /**
+     * Get all refreshed partitions' plan of the materialized view.
+     * @param mvContext: materialized view context
+     * @return:  a pair of compensated mv scan plan(refreshed partitions) and its output columns
+     */
+    private static Pair<OptExpression, List<ColumnRefOperator>> getMvScanPlan(MaterializationContext mvContext) {
+        // NOTE: mv's scan operator has already been partition pruned by filtering refreshed partitions,
+        // see MvRewritePreprocessor#createScanMvOperator.
+        final LogicalOlapScanOperator mvScanOperator = mvContext.getScanMvOperator();
+        final MaterializedView mv = mvContext.getMv();
+        final LogicalOlapScanOperator.Builder mvScanBuilder = OperatorBuilderFactory.build(mvScanOperator);
+        final List<ColumnRefOperator> orgMvScanOutputColumns = MvRewritePreprocessor.getMvScanOutputColumnRefs(mv,
+                mvScanOperator);
+        mvScanBuilder.withOperator(mvScanOperator);
+        OptExpression mvScanOptExpression = OptExpression.create(mvScanBuilder.build());
+        deriveLogicalProperty(mvScanOptExpression);
+        // duplicate mv's plan and output columns
+        OptExpressionDuplicator duplicator = new OptExpressionDuplicator(mvContext);
+        OptExpression newMvScanPlan = duplicator.duplicate(mvScanOptExpression);
+        // output columns order by mv's columns
+        List<ColumnRefOperator> mvScanOutputColumns = duplicator.getMappedColumns(orgMvScanOutputColumns);
+        newMvScanPlan.getOp().setOpRuleMask(OP_UNION_ALL_BIT);
+        return Pair.create(newMvScanPlan, mvScanOutputColumns);
+    }
+
+    /**
+     * Get all to-refreshed partitions' plan of the materialized view.
+     * @param mvContext: materialized view context
+     * @param mvCompensation: materialized view's compensation info
+     * @return:  a pair of compensated mv query scan plan(to-refreshed partitions) and its output columns
+     */
+    private static Pair<OptExpression, List<ColumnRefOperator>> getMvQueryPlan(MaterializationContext mvContext,
```

Comment:
`getMvQueryPlan ` will not consider rewrite  match type. It only use the mv's defined query to make sure the transparent plan is correctly composed by mv and mv's defined query plan.

But once it can use transparent plan which means the query can be rewritten by the mv, so use transparent mv is ok even query rewrite mode is VIEW_DELTA?

Reply:
it will incur many more table scans, this may be a big problem?

Reply:
if it is not solved, better to disable it in view delta join scene.

Reply:
Good Advise. I will handle it when we use transparent union rewrite by default.